File: Classes/Chapter.m
//
//  Chapter.m
//  AePubReader
//
//  Created by Federico Frappi on 08/04/11.
//  Copyright 2011 __MyCompanyName__. All rights reserved.
//

#import "Chapter.h"
#import "NSString+HTML.h"

@interface Chapter() {
    UIWebView *webView;
}

@end

@implementation Chapter 

- (id) initWithPath:(NSString*)theSpinePath title:(NSString*)theTitle chapterIndex:(int) theIndex{
    if((self=[super init])){
        self.spinePath = theSpinePath;
        self.title = theTitle;
        self.chapterIndex = theIndex;

		NSString *html = [[NSString alloc] initWithData:[NSData dataWithContentsOfURL:[NSURL fileURLWithPath:theSpinePath]] encoding:NSUTF8StringEncoding];
		self.text = [html stringByConvertingHTMLToPlainText];
    }
    return self;
}

- (void) loadChapterWithWindowSize:(CGRect)theWindowSize fontPercentSize:(int) theFontPercentSize{
    self.fontPercentSize = theFontPercentSize;
    self.windowSize = theWindowSize;
//  NSLog(@"webviewSize: %f * %f, fontPercentSize: %d", theWindowSize.size.width, theWindowSize.size.height,theFontPercentSize);
    webView = [[UIWebView alloc] initWithFrame:self.windowSize];
    [webView setDelegate:self];
    NSURLRequest* urlRequest = [NSURLRequest requestWithURL:[NSURL fileURLWithPath:self.spinePath]];
    [webView loadRequest:urlRequest];
}

- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error{
    NSLog(@"%@", error);
}

- (void) webViewDidFinishLoad:(UIWebView*)webView{
    NSString *varMySheet = @"var mySheet = document.styleSheets[0];";
	
	NSString *addCSSRule =  @"function addCSSRule(selector, newRule) {"
	"if (mySheet.addRule) {"
        "mySheet.addRule(selector, newRule);"								// For Internet Explorer
	"} else {"
        "ruleIndex = mySheet.cssRules.length;"
        "mySheet.insertRule(selector + '{' + newRule + ';}', ruleIndex);"   // For Firefox, Chrome, etc.
    "}"
	"}";
	
//	NSLog(@"w:%f h:%f", webView.bounds.size.width, webView.bounds.size.height);
	
	NSString *insertRule1 = [NSString stringWithFormat:@"addCSSRule('html', 'padding: 0px; height: %fpx; -webkit-column-gap: 0px; -webkit-column-width: %fpx;')", webView.frame.size.height, webView.frame.size.width];
	NSString *insertRule2 = [NSString stringWithFormat:@"addCSSRule('p', 'text-align: justify;')"];
	NSString *setTextSizeRule = [NSString stringWithFormat:@"addCSSRule('body', '-webkit-text-size-adjust: %d%%;')", self.fontPercentSize];
    
	
	[webView stringByEvaluatingJavaScriptFromString:varMySheet];
	
	[webView stringByEvaluatingJavaScriptFromString:addCSSRule];
		
	[webView stringByEvaluatingJavaScriptFromString:insertRule1];
	
	[webView stringByEvaluatingJavaScriptFromString:insertRule2];
	
    [webView stringByEvaluatingJavaScriptFromString:setTextSizeRule];
    
	int totalWidth = [[webView stringByEvaluatingJavaScriptFromString:@"document.documentElement.scrollWidth"] intValue];
	self.pageCount = (int)((float)totalWidth/webView.bounds.size.width);
	
//    NSLog(@"Chapter %d: %@ -> %d pages", chapterIndex, title, pageCount);
    
    [self.delegate chapterDidFinishLoad:self];
    
}



@end

File: SK_EPReader/Classes/Chapter.swift
//
//  Chapter.swift
//  SK_EPReader
//
//  Created by Yury Lapitsky on 2/19/16.
//  Copyright © 2016 skyylex. All rights reserved.
//

import Foundation
import UIKit

@objc protocol ChapterDelegate: NSObjectProtocol {
    func chapterDidFinishLoad(chapter: Chapter)
}

// TODO: chapter should be struct
class Chapter: NSObject, UIWebViewDelegate {
     /// TODO: rename to index
    var chapterIndex: Int = 0
    var pageCount: Int = 0
    var fontPercentSize: Int = 100
    var spinePath: String = ""
    var title: String = ""
    var text: String = ""
    
    var windowSize: CGRect = CGRect()
    
    /// TODO: remove from here, chapter needs to be PO
    var delegate: ChapterDelegate?
    
    private var webView: UIWebView?
}

extension Chapter {
    convenience init(spinePath: String, title: String, chapterIndex: Int) {
        self.init()
        
        self.spinePath = spinePath
        self.title = title
        self.chapterIndex = chapterIndex
            
        let spineData =  NSData(contentsOfURL:NSURL(fileURLWithPath: spinePath))
        /// TODO: force unwrap need to be replaced with validation before creation
        self.text = NSString(data:spineData!, encoding:NSUTF8StringEncoding) as! String
    }
    
    func load(size: CGRect, fontPercentSize: Int) {
        self.fontPercentSize = fontPercentSize
        windowSize = size
        
        webView = UIWebView(frame: windowSize)
        webView!.delegate = self
        let request = NSURLRequest(URL: NSURL(fileURLWithPath: spinePath))
        webView!.loadRequest(request)
    }
    
    func webViewDidFinishLoad(webView: UIWebView) {
        let varMySheetJS = "var mySheet = document.styleSheets[0];"
        
        var addCSSRuleJS = "function addCSSRule(selector, newRule) {"
        addCSSRuleJS += "if (mySheet.addRule) {"
        addCSSRuleJS += "mySheet.addRule(selector, newRule);"								// For Internet Explorer
        addCSSRuleJS += "} else {"
        addCSSRuleJS += "ruleIndex = mySheet.cssRules.length;"
        addCSSRuleJS += "mySheet.insertRule(selector + '{' + newRule + ';}', ruleIndex);"   // For Firefox, Chrome, etc.
        addCSSRuleJS += "}"
        addCSSRuleJS += "}"
        
        let columnSizesJS = String(format: "addCSSRule('html', 'padding: 0px; height: %fpx; -webkit-column-gap: 0px; -webkit-column-width: %fpx;')", webView.frame.height, webView.frame.width)
        let paragraphTextAlignJS = "addCSSRule('p', 'text-align: justify;')"
        let bodyTextAdjustJS = String(format: "addCSSRule('body', '-webkit-text-size-adjust: %d%%;')", fontPercentSize)
        
        webView.stringByEvaluatingJavaScriptFromString(varMySheetJS)
        webView.stringByEvaluatingJavaScriptFromString(addCSSRuleJS)
        webView.stringByEvaluatingJavaScriptFromString(columnSizesJS)
        webView.stringByEvaluatingJavaScriptFromString(paragraphTextAlignJS)
        webView.stringByEvaluatingJavaScriptFromString(bodyTextAdjustJS)
        
        let result = webView.stringByEvaluatingJavaScriptFromString("document.documentElement.scrollWidth")
        if let pageCount = Int(result!) {
            self.pageCount = (webView.bounds.size.width != 0) ? Int(CGFloat(pageCount) / webView.bounds.size.width) : 0;
            
            self.webView = nil
            
            self.delegate?.chapterDidFinishLoad(self)
        }
    }
    
    func webView(webView: UIWebView, didFailLoadWithError error: NSError?) {
        print(error)
    }
}